src/tool/cpp/cppwinrt/strings/base_await.h

```diff
@@ -671,77 +672,100 @@ namespace winrt::impl
+            }
+
+            rethrow_if_failed();
+            WINRT_ASSERT(m_status == AsyncStatus::Started);
+            throw hresult_illegal_method_call();
+        }
+
         AsyncInterface get_return_object() const noexcept
         {
             return*this;
         }
 
-        std::experimental::suspend_never initial_suspend() const noexcept
+        void get_return_value() const noexcept
         {
-            return{};
         }
 
-        struct final_suspend_type
+        void set_completed() noexcept
         {
-            promise_base* promise;
-
-            bool await_ready() const noexcept
-            {
-                return false;
-            }
-
-            void await_resume() const noexcept
-            {
-            }
+            CompletedHandler handler;
+            AsyncStatus status;
 
-            bool await_suspend(std::experimental::coroutine_handle<>) const noexcept
             {
-                uint32_t const remaining = promise->subtract_reference();
+                slim_lock_guard const guard(m_lock);
 
-                if (remaining == 0)
+                if (m_status == AsyncStatus::Started)
                 {
-                    std::atomic_thread_fence(std::memory_order_acquire);
+                    m_status = AsyncStatus::Completed;
                 }
 
-                return remaining > 0;
+                handler = std::move(this->m_completed);
+                status = this->m_status;
             }
-        };
 
-        final_suspend_type final_suspend() noexcept
-        {
-            return{ this };
+            if (handler)
+            {
+                handler(*this, status);
+            }
         }
 
-        void unhandled_exception() noexcept
+        std::experimental::suspend_never initial_suspend() const noexcept
         {
-            CompletedHandler handler;
-            AsyncStatus status;
+            return{};
+        }
 
+        auto final_suspend() noexcept
+        {
+            struct awaiter
             {
-                slim_lock_guard const guard(m_lock);
-                WINRT_ASSERT(m_status == AsyncStatus::Started || m_status == AsyncStatus::Canceled);
-                m_exception = std::current_exception();
+                promise_base* promise;
 
-                try
+                bool await_ready() const noexcept
                 {
-                    std::rethrow_exception(m_exception);
+                    return false;
                 }
-                catch (hresult_canceled const&)
+
+                void await_resume() const noexcept
                 {
-                    m_status = AsyncStatus::Canceled;
                 }
-                catch (...)
+
+                bool await_suspend(std::experimental::coroutine_handle<>) const noexcept
                 {
-                    m_status = AsyncStatus::Error;
+                    uint32_t const remaining = promise->subtract_reference();
```

Comment:
Should `subtract_reference` be after `promise->set_completed()`?
As soon as I gave up my reference, I assume that the promise may be destroyed by someone else concurrently which decremented its own reference and noticed that the count went to zero.